### PR TITLE
QPOptions

### PR DESCRIPTION
**Pull request by davidhadas**
"Fixes [#5170](https://github.com/knative/docs/issues/5170)

## Proposed Changes 

- Add documentation about Queue Proxy Extensions feature merge into Queue Proxy

### CODE REVIEW

**Comment by abrennan89** [comment]
It looks like at https://pkg.go.dev/knative.dev/security-guard/pkg/qpoption this is actually named "Queue Proxy Option" rather than "Queue Proxy Extension". Please rename it across the doc so that it's consistent, i.e. use Queue Proxy Option since that's the actual feature name, otherwise this could cause confusion.

File: docs/serving/configuration/feature-flags.md
Comment (abrennan89) [reply docs/serving/configuration/feature-flags.md:430]: same comment also re literal name of pod info volume?

File: docs/serving/queue-extensions.md
Comment (abrennan89) [inline docs/serving/queue-extensions.md:1]: this title is not really descriptive; it doesn't tell a user why they would want to read this section or what information to expect

File: docs/serving/queue-extensions.md
Comment (abrennan89) [inline docs/serving/queue-extensions.md:3]: doesn't explain what Queue Proxy is or why someone would want to extend it?

Comment (abrennan89) [inline docs/serving/queue-extensions.md:3]:
```suggestion
Queue Proxy can be built with additional runtime extensions enabled that provide optional features.
```
I'd simplify this by a lot and instead explain what Queue Proxy is and why it matters in Knative. The stuff about downstream etc should be explained by different vendors in their downstream docs instead, i.e. they can call out which additional features if any are available in their offering. For core Knative docs we should instead explain the feature and what the possible options are.

File: docs/serving/queue-extensions.md
Comment (abrennan89) [inline docs/serving/queue-extensions.md:5]:
```suggestion
For example, the [`security-guard`](https://knative.dev/security-guard) repository provides a Queue Proxy Option [qpoption](https://knative.dev/security-guard/pkg/qpoption) package, which enables users to add additional security features to Queue Proxy.
```

File: docs/serving/queue-extensions.md
Comment (abrennan89) [inline docs/serving/queue-extensions.md:10]:
```suggestion
The runtime features available are determined when the Queue Proxy image is built. Queue Proxy defines an orderly manner to activate and to configure extensions.

#### Additional information

- [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of Extensions.
- [Using Queue Proxy Extensions](./services/using-queue-extensions.md) - discussing how to configure a service to use features implemented in extensions.
```

File: docs/serving/queue-extensions.md
Comment (abrennan89) [inline docs/serving/queue-extensions.md:12]:
```suggestion
#### Adding extentions

You can add extensions by replacing the `cmd/queue/main.go` file before the Queue Proxy image is built. The following example shows a `cmd/queue/main.go` file that adds the `test-gate` extension:
```

**Review by abrennan89 — CHANGES_REQUESTED** [review]
github acting up for me, will add more comments for rest soon

File: docs/serving/services/using-queue-extensions.md
Comment (abrennan89) [inline docs/serving/services/using-queue-extensions.md:3]: Please refer to the docs style guide re formatting for things like Services https://github.com/knative/docs/blob/main/contribute-to-docs/style-guide/style-and-formatting.md

File: docs/serving/services/using-queue-extensions.md
Comment (abrennan89) [inline docs/serving/services/using-queue-extensions.md:12]: same comment re formatting for Service

Comment (abrennan89) [inline docs/serving/services/using-queue-extensions.md:12]: re queue proxy extensions use capitalization and naming consistently throughout docs please

**Review by abrennan89 — CHANGES_REQUESTED** [review]

**Comment by abrennan89** [comment]
@davidhadas left some comments for now, will provide a second review once some of these are fixed

/hold

**Review by davidhadas — COMMENTED** [review]
Send a new commit with changes that hopefully capture all comments

File: config/nav.yml
Comment (snneji) [inline config/nav.yml:106]:
```suggestion
          - Using extensions enabled by QPoptions: serving/services/using-queue-extensions.md
```

Comment (snneji) [reply config/nav.yml:106]: Make the name similar to the other items

Comment (snneji) [reply config/nav.yml:106]: I'm wondering whether it would be better to say "Queue Proxy options" instead of "QPoptions"? 

Also, I've seen "QPOptions" and "QPoptions". Make sure one or the other is consistently used.

File: docs/serving/queue-extensions.md
Comment (snneji) [inline docs/serving/queue-extensions.md:8]:
```suggestion
You can extend Queue Proxy to offer additional features. The QPOptions feature of Queue Proxy allows additional runtime packages to extend Queue Proxy capabilities.
```

Comment (snneji) [reply docs/serving/queue-extensions.md:8]: Should it be "QPOptions" or "QPoptions" here? Make sure one or the other is consistently used.

File: docs/serving/queue-extensions.md
Comment (snneji) [inline docs/serving/queue-extensions.md:10]: Check "QPOptions" vs "QPoptions" here also.

File: docs/serving/queue-extensions.md
Comment (snneji) [inline docs/serving/queue-extensions.md:16]:
```suggestion
- [Enabling Queue Proxy Pod Info](./configuration/feature-flags.md#queue-proxy-pod-info) - discussing a necessary step to enable the use of extensions.
```

File: docs/serving/queue-extensions.md
Comment (snneji) [inline docs/serving/queue-extensions.md:14]:
```suggestion
## Additional information
```

Comment (snneji) [reply docs/serving/queue-extensions.md:14]: Don't use an H4 heading that's not preceded by an H3.

File: docs/serving/queue-extensions.md
Comment (snneji) [inline docs/serving/queue-extensions.md:19]:
```suggestion
## Adding extensions
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:5]: Probably a good idea to move this section down so that the prerequisites are grouped together.

The Prereqs could be in one section, for example: 

```
## Prerequisites

Before you can use extensions enabled by QPoptions, you must:

- Prepare your cluster:
  - Make sure you are using a Queue Proxy image...
  - Make sure that the cluster config-features is set with...
- Meet the prerequisites in [Creating a Service](...)
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:7]:
```suggestion
1. Make sure you are using a Queue Proxy image that was built with the extensions that you wish to use. See [Extending Queue Proxy image with QPoptions](../queue-extensions.md).
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:8]:
```suggestion
1. Make sure that the cluster config-features is set with `queueproxy.mount-podinfo: allowed`. See [Enabling Queue Proxy Pod Info](../configuration/feature-flags.md#queue-proxy-pod-info) for more details.
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:14]:
```suggestion
Setting a value of: `qpoption.knative.dev/<ExtensionName>-activate: "enable"` activates the extension.
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:16]:
```suggestion
Setting a value of: `qpoption.knative.dev/<extension-name>-config-<key>: "<value>"` adds a configuration of `key: value` to the extension.
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:18]:
```suggestion
In addition, the Service must ensure that the Pod Info volume is mounted by adding the `features.knative.dev/queueproxy-podinfo: enabled` annotation under the: `spec.template.metadata` of the Service CRD.
```

File: docs/serving/services/using-queue-extensions.md
Comment (snneji) [inline docs/serving/services/using-queue-extensions.md:80]:
```suggestion
After the Service has been created, Knative propagates the annotations to the podSpec of the Service deployment. When a Service pod is created, the Queue Proxy sidecar will mount a volume that contains the pod annotations and activate the `testgate` extension. This occurs if the `testgate` extension is available in the Queue Proxy image. The `testgate` extension will then be configured with the configuration: `{ sender: "Joe", response: "CU"}`.
```

**Review by snneji — COMMENTED** [review]
I've left a few suggestions, but I'm not sure if @abrennan89 will also want to do a second review as stated in her previous comment.

File: docs/serving/queue-extensions.md
Comment (davidhadas) [reply docs/serving/queue-extensions.md:10]: will do a full search and align in a separate commit

File: config/nav.yml
Comment (davidhadas) [reply config/nav.yml:106]: I will align all to QPOptions

Comment (davidhadas) [reply config/nav.yml:106]: replacing to Using extensions enabled by QPoptions need to be aligned with links  - in a separate commit

File: docs/serving/services/using-queue-extensions.md
Comment (abrennan89) [inline docs/serving/services/using-queue-extensions.md:3]: It still doesn't explain what QPOptions is here - I would like to see a proper conceptual intro so that users know why they would enable this or what they would use it for. For example, something like:

"QPOptions is an optional Go package that provides extensions for Knative Services. These extensions allow you to..." etc

I don't think what's currently here is enough to warrant adding a new section to the docs.

**Comment by snneji** [comment]
Looks good to me. @abrennan89 will do her second review and then we can get this merged. 🙂

**Comment by abrennan89** [comment]
@snneji I'll leave it to you to merge this one if you're happy with it

File: docs/serving/services/using-queue-extensions.md
Comment (davidhadas) [reply docs/serving/services/using-queue-extensions.md:5]: Added an intro to what is QPOptions

**Pull request merged**
